include/tvm/layout.h

```diff
@@ -0,0 +1,435 @@
+  static const LayoutAxis r;
+  static const LayoutAxis s;
+  static const LayoutAxis t;
+  static const LayoutAxis u;
+  static const LayoutAxis v;
+  static const LayoutAxis w;
+  static const LayoutAxis x;
+  static const LayoutAxis y;
+  static const LayoutAxis z;
+
+  static const LayoutAxis& Get(const char name) {
+    switch (name) {
+      case 'A': return A;
+      case 'B': return B;
+      case 'C': return C;
+      case 'D': return D;
+      case 'E': return E;
+      case 'F': return F;
+      case 'G': return G;
+      case 'H': return H;
+      case 'I': return I;
+      case 'J': return J;
+      case 'K': return K;
+      case 'L': return L;
+      case 'M': return M;
+      case 'N': return N;
+      case 'O': return O;
+      case 'P': return P;
+      case 'Q': return Q;
+      case 'R': return R;
+      case 'S': return S;
+      case 'T': return T;
+      case 'U': return U;
+      case 'V': return V;
+      case 'W': return W;
+      case 'X': return X;
+      case 'Y': return Y;
+      case 'Z': return Z;
+      case 'a': return a;
+      case 'b': return b;
+      case 'c': return c;
+      case 'd': return d;
+      case 'e': return e;
+      case 'f': return f;
+      case 'g': return g;
+      case 'h': return h;
+      case 'i': return i;
+      case 'j': return j;
+      case 'k': return k;
+      case 'l': return l;
+      case 'm': return m;
+      case 'n': return n;
+      case 'o': return o;
+      case 'p': return p;
+      case 'q': return q;
+      case 'r': return r;
+      case 's': return s;
+      case 't': return t;
+      case 'u': return u;
+      case 'v': return v;
+      case 'w': return w;
+      case 'x': return x;
+      case 'y': return y;
+      case 'z': return z;
+      default: LOG(FATAL) << "Invalid layout axis name " << name;
+    }
+    // suppress return-type warning.
+    return A;
+  }
+
+  // Get the singleton LayoutAxis using itvar->var->name_hint
+  inline static const LayoutAxis& Get(const IterVar& itvar) {
+    const std::string axis = itvar->var.get()->name_hint;
+    CHECK_EQ(axis.size(), 1) << "Invalid layout axis " << axis;
+    return LayoutAxis::Get(axis[0]);
+  }
+
+  // Get the singleton LayoutAxis using name[0] (size of name must be 1).
+  inline static const LayoutAxis& make(const std::string& name) {
+    CHECK_EQ(name.length(), 1) << "Invalid axis " << name;
+    return LayoutAxis::Get(name[0]);
+  }
+
+  inline bool IsPrimal() const { return name_ >= 'A' && name_ <= 'Z'; }
+  inline std::string name() const { return name_str_; }
+
+  // if current axis is primal, switch the axis to its subordinate one,
+  // else switch to the primal.
+  inline const LayoutAxis& to_dual() const {
+    if (name_ >= 'A' && name_ <= 'Z') {
+      return LayoutAxis::Get(name_ - 'A' + 'a');
+    } else {
+      return LayoutAxis::Get(name_ - 'a' + 'A');
+    }
+  }
+
+  // return the primal axis. If it is already primal, return itself.
+  inline const LayoutAxis& to_primal() const {
```

Comment:
Style: Use CamelCase

we don't need to explicit mark inline in the function body